stetho/src/main/java/com/facebook/stetho/common/android/HandlerUtil.java

```diff
@@ -11,38 +11,104 @@
 
 import android.os.Handler;
 import android.os.Looper;
+import com.facebook.stetho.common.Util;
+
+import java.util.concurrent.Callable;
 
 public final class HandlerUtil {
   private HandlerUtil() {
   }
 
-  public static boolean postAndWait(Handler handler, final Runnable r) {
-    if (Looper.myLooper() == handler.getLooper()) {
-      r.run();
-      return true;
+  /**
+   * Checks whether the current thread is the same thread that the {@link Handler} is associated
+   * with.
+   * @return true if the current thread is the same thread that the {@link Handler} is associated
+   * with; otherwise false.
+   */
+  public static boolean checkAccess(Handler handler) {
+    return Looper.myLooper() == handler.getLooper();
+  }
+
+  /**
+   * Enforces that the current thread is the same thread that the {@link Handler} is associated
+   * with.
+   * @throws IllegalStateException if the current thread is not the same thread that the
+   * {@link Handler} is associated with.
+   */
+  public static void verifyAccess(Handler handler) {
+    if (!checkAccess(handler)) {
+      throw new IllegalStateException();
+    }
+  }
+
+  /**
+   * Synchronously executes a {@link Callable} on the {@link Handler} and returns its result.
+   * @param handler the {@link Handler} to use
+   * @param c the {@link Callable} to execute
+   * @param <V> the return type of the {@link Callable}
+   * @return the return value from {@link Callable#call()}
+   * @throws RuntimeException if the {@link Callable} could not be executed (e.g.
+   * {@link Handler#post(Runnable)} returned false}, or if {@link Callable#call()} threw an
+   * exception (in which case the cause will be the actual exception that was thrown).
+   */
+  public static <V> V callAndWait(Handler handler, Callable<V> c) {
+    if (checkAccess(handler)) {
+      try {
+        return c.call();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    WaitableCallable<V> wrapper = new WaitableCallable<>(c);
+    if (!handler.post(wrapper)) {
+      throw new RuntimeException("Handler.post() returned false");
+    }
+
+    wrapper.join();
+    if (wrapper.isException()) {
+      throw new RuntimeException(wrapper.getException());
+    }
+
+    return wrapper.getResult();
```

Comment:
Prefer a more concise utility that does all three of these things:
```java
getResultAndJoin();
```
It can throw as well.